### DISCUSSION TITLE
General questions on current project status and how to help

### DISCUSSION THREAD

**Discussion by miguelpduarte** [taskchampion/taskchampion#349] [Q&A] [answered]
Hello! :wave: 

This project caught my attention after seeing it mentioned in the Taskwarrior IRC.

I have some questions, as the project seems quite interesting and I'm seeing how I can help:
* Syncing: Does the syncing mechanism work with taskwarrior? I'm guessing this is not the case since it seems like a completely different model. In this case, are there plans to attempt to "migrate" the work done on the syncing part to taskwarrior? It would be quite groundbreaking, as reliable syncing seems to be taskwarrior's biggest pain point (at least in my experience).
* Would it be "worth it" to work on the CLI interface or will this be integrated in taskwarrior (follow-up of the previous question) or another tool. I've seen in some discussions the idea that making the CLI like taskwarrior may be a big challenge seeing as it's quite polished (and "merging" the two projects may happen eventually?), but I feel like having an intuitive interface will be a great selling point for adoption of the project.
* How can I/someone looking to contribute to the project help the most at the moment? Is it more relevant to work on the CLI? Start tackling open issues? Focus on developing tools that integrate with taskchampion? Invest in improving github action pipelines? Create packages for package managers to make the project easier to use? Eventually try to create a mobile app?
* Also about contribution: The good first issue label is currently assigned to 5 CLI-scoped issues. Given that these are from a while ago (late 2020), are these still relevant? I found them while looking into how to get my hands dirty.

I don't mean to impose by asking all these questions (as I understand that this may require effort to address - especially issue management, or finding work to assign to new contributors), but since I really enjoy the whole concept behind taskwarrior and have been burned several times with some of the issues it has, this seems like an interesting project to contribute to, and I want to help out.  I am also learning Rust (and am enjoying it a lot) so this seems like a great opportunity to put what I have been learning to the test :)

**Edit:** I have since found https://github.com/taskchampion/taskchampion/issues/52 which gives a recent(?) update on the state of things. Still unsure how to best help, and would appreciate any light that can be shed on this :)

**Comment by djmitche** [comment, marked as answer]
Hey, welcome!

TaskChampion has been through a few significant changes in direction, so yeah -- things are a bit confusing.  Sorry about that!

No, TC and TW have completely different sync backends.

At the moment, my plan is to integrate TC into TW as the "core" -- handling task storage and sync -- with the existing CLI support remaining in place.  That basically means the `taskchampion` and taskchampion-lib` crates, and not the CLI crate.  That said, once that merger is complete, it'd be possible to sync between the TC CLI and the TW CLI, so it might be useful anyway!  What I'd say right now is this: the CLI is not currently on the "critical path" that I'm working on, but you're right that having a nice interface (without some of the historical baggage of TW's CLI) would be nice.  So, if it's interesting code for you to work on, and you're OK not being on that "critical path", then by all means I'm happy to collaborate with you on making improvements there!

I think the "critical path" bits right now are best identified as [the 0.5.0 milestone](https://github.com/taskchampion/taskchampion/milestone/2).  #384 in particular might be a nice first bug.  If you can work on that, make a comment and I'll assign it to you -- it's only assigned to me because I intended to do it soon (and assigned issues show up in my taskwarrior task list!).

In general, the bits that need to be attended to soon are:
 * getting the TW and TC data models to match (most of the 0.5.0 issues fall into this category)
 * documenting that data model and the new sync model (currently blocked on https://github.com/GothenburgBitFactory/taskwarrior/pull/2742 and https://github.com/GothenburgBitFactory/tw.org/pull/303)
 * making the TC sync server a bit nicer and easier to use/configure
 * figuring out how to set up the build process to link TC into TW (I have some work in progress [here](https://github.com/djmitche/taskwarrior/commits/use-taskchampion) but it's hard-coding paths in my home directory so not very portable!)
 * interfacing TW and TC (also work-in-progress)

Does any of that sound particularly interesting?  The server has some web-related Rust, the interface to TW involves a bunch of Rust FFI and C++, the CLI involves lots of `nom` parsers, and the docs bit involve, I think, some social engineering to get things in motion on the TW project.  And (since I see you're in infosec) there's some good thinking to be done about the privacy and security aspects of the sync-server model.

**Reply by miguelpduarte** [reply]
Thanks for the detailed writeup! I figured out some of the current status after finding recent comments in #52. Would it be relevant to have a pinned issue or discussion that would be a sort of "megathread" for any newcomer that wants to see how the project is doing and how they can help? I don't want to divert bandwidth from more critical issues, though.

In terms of issues, I was actually looking at #348 and #334 for potential first issues, as I'm still getting familiar with the codebase. I'm willing to take on the challenge for solving either of them (maybe #348 since you suggested it). However, I still have to take a bit to read the code, configure the project locally, so I might take a bit to get started. If that's okay I would be glad to take the issue, though! :slightly_smiling_face: 

On the "upcoming tasks": All of them sound quite interesting, to be honest, as they both focus on things I've seen as well as things that are new to me, which is awesome. I'm not sure what would be simplest to start with, but maybe some "investigative work" (from an outsider pov) of the TC sync server setup would be a good first step?  
Once again, everything sounds cool, so if there are higher priorities I can also focus on that :)

I'm glad to contribute in any way I can, since I really do love the whole concept of a CLI task management tool with so much customizability and filtering, and this integration really feels like an amazing leap for taskwarrior. Kudos!

**Reply by djmitche** [reply]
Sounds good!  Yes, a pinned discussion would be a good idea, if that's possible.  I'll put that on my task list ;)

There's no great rush, I think. When I get to integrating task modification in my work on TaskWarrior, I'll need #348, but that's probably a few weeks away at best.  (also: commenting on the issue is actually a crucial step as it's what adds you to the list of possible assignees in the GitHub UI)

And yes, investigating the server side of things seems like a good choice.  We did notice earlier that Actix seems to be overkill for this particular usage, and maybe just using Hyper would be better, so that's something to consider.  Please ask questions, too -- I'm happy to help :)

**Reply by miguelpduarte** [reply]
Thanks once again for the feedback! I commented on the issue and am starting to look through the code more thoroughly. I'll also try to document my process / questions so that pain points in contributing to TC can be possibly be found and addressed. I'll keep you posted :)
